plugins/field-date/src/field_date.ts

```diff
@@ -0,0 +1,263 @@
+   */
+  protected override doClassValidation_(newValue?: string): string | null {
+    const newDate = typeof newValue === 'string' ? new Date(newValue) : null;
+    if (!newDate || isNaN(newDate.getTime())) return null;
+
+    // NOTE: Although `newValue` is already in the correct format for
+    // `FieldDate` here (yyyy-mm-dd), convert `newDate` anyway to allow
+    // subclasses to try and save any date-compatible strings and correct back
+    // to yyyy-mm-dd format.
+    return getHTMLInputDateValue(newDate);
+  }
+
+  /**
+   * Get the text to display on the block when the input hasn't spawned in.
+   *
+   * @returns The text to display on the block.
+   */
+  protected override getText_(): string | null {
+    const value = this.getValue();
+    if (!value) return null;
+    // NOTE: There may be discrepancies between the text and the input based on 
+    // browser. For example, 'en-US' would display the text '2/14/2020', then 
+    // clicking in Safari on iOS would display 'Feb 14, 2020'.
+
+    // TODO: Should we handle these distinctions? Anytime browser format 
+    // updated, we would need to update here.
+    return getLocaleDateString(value);
+  }
+
+  /**
+   * Renders the field. If the picker is shown make sure it has the current
+   * date selected.
+   */
+  protected render_() {
+    super.render_();
+  }
+
+  /**
+   * Shows the inline free-text editor on top of the text along with the date
+   * editor.
+   * @param e Optional mouse event that triggered the field to
+   *     open, or undefined if triggered programmatically.
+   * @param _quietInput Quiet input.
+   * @override
+   */
+  protected showEditor_(e?: Event, _quietInput?: boolean) {
+    if (this.sourceBlock_) {
+      // NOTE: Disable modal inputs to use default browser inputs on mobile.
+      this.sourceBlock_.workspace.options.modalInputs = false;
+    }
+
+    super.showEditor_(e, _quietInput);
+    this.showDropdown();
+  }
+
+  /**
+   * Updates the size of the field based on the text.
+   *
+   * @param margin margin to use when positioning the text element.
+   */
+  protected override updateSize_(margin?: number) {
+    // Add margin so that the date input's datepicker icon doesn't clip with 
+    // the text when sized for the date.
+    super.updateSize_((margin ?? 0) + 20);
+  }
+
+  /**
+   * Shows the datepicker.
+   */
+  private showDropdown(): void {
+    if (!this.htmlInput_) return;
+    Blockly.utils.dom.addClass(this.htmlInput_!, 'blocklyDateInput');
+
+    // NOTE: HTMLInputElement.showPicker() is not available in earlier
+    // TypeScript versions (like 4.7.4), so casting to `any` to be compatible
+    // with dev scripts. Additionally, it's not available for date inputs for 
+    // Safari. For browser compatibility of showPicker, see:
+    // https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/showPicker
+    /* eslint-disable @typescript-eslint/no-explicit-any */
+    (this.htmlInput_ as any).showPicker();
```

Comment:
You might be able to remove this `any` cast b/c we updated dev scripts.

Reply:
Since the TypeScript rule is "^4.3.2 || ^5" and the package-lock for dev-scripts set to `4.7.4`, unfortunately I can't without another update to dev-scripts.